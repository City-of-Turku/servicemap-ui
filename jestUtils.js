import React from 'react';
import { render } from '@testing-library/react';
import configureStore from 'redux-mock-store';
import { ThemeProvider } from '@mui/material/styles';
import { IntlProvider } from 'react-intl';
import { Provider } from 'react-redux';
import themes from './src/themes';
import finnish from './src/i18n/fi';

const mockStore = configureStore([]);

// Mock props for intl provider
const intlMock = {
  locale: 'fi',
  messages: finnish,
  wrapRichTextChunksInFragment: false,
};

const Providers = (mockState) => ({ children }) => {
  const store = mockStore(mockState);
  return (
    <Provider store={store}>
      <IntlProvider {...intlMock}>
        <ThemeProvider theme={themes.SMTheme}>
          {children}
        </ThemeProvider>
      </IntlProvider>
    </Provider>
  );
}

export const getRenderWithProviders = mockState => component => render(component, { wrapper: Providers(mockState) });