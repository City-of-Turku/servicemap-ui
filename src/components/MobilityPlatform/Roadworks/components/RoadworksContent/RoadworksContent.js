import { Typography } from '@mui/material';
import React from 'react';
import PropTypes from 'prop-types';
import styled from '@emotion/styled';
import { format } from 'date-fns';
import { StyledContainer, StyledHeaderContainer, StyledTextContainer } from '../../../styled/styled';

const RoadworksContent = ({ item }) => {
  const roadworkDetails = item?.announcements[0];

  const formatDate = dateTimeValue => format(new Date(dateTimeValue), 'dd.MM.yyyy');

  const filterRestrictions = restrictionsData => {
    const restrictionTypes = ['SPEED_LIMIT', 'SPEED_LIMIT_LENGTH'];
    if (restrictionsData?.length > 0) {
      return restrictionsData.filter(restriction => restrictionTypes.includes(restriction.type));
    }
    return [];
  };

  const roadWorksRestrictions = filterRestrictions(roadworkDetails.additional_info?.restrictions);

  const renderRestrictions = () => (
    roadWorksRestrictions?.length > 0 ? (
      roadWorksRestrictions.map(limitItem => (
        <StyledTextContainer key={limitItem.restriction.quantity}>
          <Typography variant="body2">
            {`${limitItem.restriction.name}: ${limitItem.restriction.quantity} ${limitItem.restriction.unit}`}
          </Typography>
        </StyledTextContainer>
      ))
    ) : null
  );

  const renderExtraFeatures = () => (
    roadworkDetails.features?.length > 0 ? (
      roadworkDetails.features.map(feature => (
        <React.Fragment key={feature.name}>
          {feature.quantity && feature.unit ? (
            <StyledTextContainer>
              <Typography variant="body2">
                {`${feature.name}: ${feature.quantity} ${feature.unit}`}
              </Typography>
            </StyledTextContainer>
          ) : (
            <StyledTextContainer>
              <Typography variant="body2">
                {feature.name}
              </Typography>
            </StyledTextContainer>
          )}
        </React.Fragment>
      ))
    ) : null
  );

  const renderDateValues = () => {
    if (roadworkDetails?.additional_info?.timeAndDuration?.startTime && roadworkDetails?.additional_info.timeAndDuration.endTime) {
      return (
        <StyledTextContainer>
          <StyledText variant="body2">
            {`Aika: ${formatDate(roadworkDetails?.additional_info?.timeAndDuration?.startTime)} - ${formatDate(
              roadworkDetails?.additional_info?.timeAndDuration?.endTime,
            )}`}
          </StyledText>
        </StyledTextContainer>
      );
    }
    if (roadworkDetails?.additional_info?.timeAndDuration?.startTime) {
      return (
        <StyledTextContainer>
          <StyledText variant="body2">
            {`Ty√∂ alkoi: ${formatDate(roadworkDetails?.additional_info?.timeAndDuration?.startTime)}`}
          </StyledText>
        </StyledTextContainer>
      );
    }
    return null;
  };

  return (
    <StyledContainer>
      <StyledHeaderContainer>
        <StyledText variant="subtitle2" component="h4">
          {roadworkDetails?.title}
        </StyledText>
      </StyledHeaderContainer>
      <div>
        <StyledTextContainer>
          <StyledText variant="body2">{roadworkDetails?.description}</StyledText>
        </StyledTextContainer>
        {roadworkDetails.comment ? (
          <StyledTextContainer>
            <StyledText variant="body2">{roadworkDetails?.comment}</StyledText>
          </StyledTextContainer>
        ) : null}
        {renderRestrictions()}
        {renderExtraFeatures()}
        {renderDateValues()}
      </div>
    </StyledContainer>
  );
};

const StyledText = styled(Typography)(({ theme }) => ({
  marginBottom: theme.spacing(0.5),
}));

RoadworksContent.propTypes = {
  item: PropTypes.shape({
    situationType: PropTypes.string,
    announcements: PropTypes.arrayOf(
      PropTypes.shape({
        title: PropTypes.string,
        description: PropTypes.string,
        comment: PropTypes.string,
        additional_info: PropTypes.shape({
          timeAndDuration: PropTypes.shape({
            startTime: PropTypes.string,
            endTime: PropTypes.string,
          }),
        }),
      }),
    ),
  }),
};

RoadworksContent.defaultProps = {
  item: {},
};

export default RoadworksContent;
