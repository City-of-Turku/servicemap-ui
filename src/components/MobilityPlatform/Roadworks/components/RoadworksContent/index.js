import RoadworksContent from './RoadworksContent';

export default RoadworksContent;
