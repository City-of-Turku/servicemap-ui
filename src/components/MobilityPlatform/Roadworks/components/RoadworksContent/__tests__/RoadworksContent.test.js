/* eslint-disable max-len */
// Link.react.test.js
import React from 'react';
import { getRenderWithProviders } from '../../../../../../../jestUtils';
import { initialState } from '../../../../../../redux/reducers/user';
import RoadworksContent from '../index';

const mockProps = {
  item: {
    announcements: [
      {
        title: 'Test roadwork',
        description: 'Test desc',
        comment: 'Comment',
        additional_info: {
          restrictions: [
            {
              type: 'SPEED_LIMIT',
              restriction: {
                name: 'Test limit',
                quantity: '50',
                unit: 'km/h',
              },
            },
          ],
          timeAndDuration: {
            startTime: '2024-04-10T10:00:00z',
            endTime: '2024-06-05T17:00:00z',
          },
        },
      },
    ],
  },
};

const renderWithProviders = getRenderWithProviders({
  user: initialState,
});

describe('<RoadworksContent />', () => {
  it('does match snapshot', () => {
    const { container } = renderWithProviders(<RoadworksContent {...mockProps} />);
    expect(container).toMatchSnapshot();
  });

  it('does show text correctly', () => {
    const { container } = renderWithProviders(<RoadworksContent {...mockProps} />);

    const h4 = container.querySelectorAll('h4');
    const p = container.querySelectorAll('p');
    const roadworkDetails = mockProps.item.announcements[0];
    const restrictionsArray = roadworkDetails.additional_info.restrictions[0];
    expect(h4[0].textContent).toContain(roadworkDetails.title);
    expect(p[0].textContent).toContain(roadworkDetails.description);
    expect(p[1].textContent).toContain(roadworkDetails.comment);
    expect(p[2].textContent).toContain(
      `${restrictionsArray.restriction.name}: ${restrictionsArray.restriction.quantity} ${restrictionsArray.restriction.unit}`,
    );
    expect(p[3].textContent).toEqual('Aika: 10.04.2024 - 05.06.2024');
  });
});
