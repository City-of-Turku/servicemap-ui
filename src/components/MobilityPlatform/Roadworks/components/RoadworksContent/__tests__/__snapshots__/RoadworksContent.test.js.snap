// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RoadworksContent /> does match snapshot 1`] = `
<div>
  <div
    class="css-zimvk6"
  >
    <div
      class="css-19od9cu"
    >
      <h4
        class="MuiTypography-root MuiTypography-subtitle2 css-wfeqti-MuiTypography-root"
      >
        Test roadwork
      </h4>
    </div>
    <div>
      <div
        class="css-jmbq56"
      >
        <p
          class="MuiTypography-root MuiTypography-body2 css-qabfq1-MuiTypography-root"
        >
          Test desc
        </p>
      </div>
      <div
        class="css-jmbq56"
      >
        <p
          class="MuiTypography-root MuiTypography-body2 css-qabfq1-MuiTypography-root"
        >
          Comment
        </p>
      </div>
      <div
        class="css-jmbq56"
      >
        <p
          class="MuiTypography-root MuiTypography-body2 css-1ak20dk-MuiTypography-root"
        >
          Test limit: 50 km/h
        </p>
      </div>
      <div
        class="css-jmbq56"
      >
        <p
          class="MuiTypography-root MuiTypography-body2 css-qabfq1-MuiTypography-root"
        >
          Aika: 05.11.2023 - 05.12.2023
        </p>
      </div>
    </div>
  </div>
</div>
`;
